packages/@aws-cdk/integ-tests/lib/assertions/providers/lambda-handler/assertion.ts

```diff
@@ -1,34 +1,151 @@
+          status: 'fail',
+          message: [
+            ...matchResult.toHumanStrings(),
+            JSON.stringify(matchResult.target, undefined, 2),
+          ].join('\n'),
+        }),
+      };
+      if (request.reportFailure) {
+        throw new Error(result.data);
+      }
+    } else {
+      result = {
+        data: JSON.stringify({
+          status: 'pass',
+        }),
+      };
+    }
+
+    return result;
+  }
+}
+
+
+class MatchCreator {
+  private readonly type: 'arrayWith' | 'objectLike' | 'exact' | 'stringLikeRegexp';
+  private readonly parsedObj: any;
+  constructor(obj: { [key: string]: any }) {
+    switch (Object.keys(obj)[0]) {
+      case '$ObjectLike':
+        this.type = 'objectLike';
+        this.parsedObj = obj.$ObjectLike;
+        break;
+      case '$ArrayWith':
+        this.type = 'arrayWith';
+        this.parsedObj = obj.$ArrayWith;
+        break;
+      case '$Exact':
+        this.type = 'exact';
+        this.parsedObj = obj.$Exact;
+        break;
+      case '$StringLike':
+        this.type = 'stringLikeRegexp';
+        this.parsedObj = obj.$StringLike;
         break;
       default:
-        throw new Error(`Unsupported query type ${request.assertionType}`);
+        this.type = 'exact';
+        this.parsedObj = obj;
     }
+  }
 
-    return result;
+  /**
+   * Return a Matcher that can be tested against the actual results.
+   * This will convert the encoded matchers into their corresponding
+   * assertions matcher.
+   *
+   * For example:
+   *
+   * ExpectedResult.objectLike({
+   *   Messages: [{
+   *     Body: Match.objectLike({
+   *       Elements: Match.arrayWith([{ Asdf: 3 }]),
+   *     }),
+   *   }],
+   * });
+   *
+   * Will be encoded as:
+   * {
+   *   $ObjectLike: {
+   *     Messages: [{
+   *       Body: {
+   *         $ObjectLike: {
+   *           Elements: {
+   *             $ArrayWith: [{ Asdf: 3 }],
+   *           },
+   *         },
+   *       },
+   *     }],
+   *   },
+   * }
+   *
+   * Which can then be parsed by this function. For each key (recursively)
+   * the parser will check if the value has one of the encoded matchers as a key
+   * and if so, it will set the value as the Matcher. So,
+   *
+   * {
+   *   Body: {
+   *     $ObjectLike: {
+   *       Elements: {
+   *         $ArrayWith: [{ Asdf: 3 }],
+   *       },
+   *     },
+   *   },
+   * }
+   *
+   * Will be converted to
+   * {
+   *   Body: Match.objectLike({
+   *     Elements: Match.arrayWith([{ Asdf: 3 }]),
+   *   }),
+   * }
+   */
+  public getMatcher(): Matcher {
+    try {
+      const final = JSON.parse(JSON.stringify(this.parsedObj), function(_k, v) {
+        const nested = Object.keys(v)[0];
+        switch (nested) {
+          case '$ArrayWith':
+            return Match.arrayWith(v[nested]);
+          case '$ObjectLike':
+            return Match.objectLike(v[nested]);
+          case '$StringLike':
+            return Match.stringLikeRegexp(v[nested]);
```

Comment:
Why is this `case` statement duplicated in the constructor? We shouldn't need both.